src/banman.cpp

```diff
@@ -0,0 +1,202 @@
+
+        LogPrint(BCLog::NET, "Loaded %d banned node ips/subnets from banlist.dat  %dms\n",
+            banmap.size(), GetTimeMillis() - nStart);
+    } else {
+        LogPrintf("Invalid or missing banlist.dat; recreating\n");
+        SetBannedSetDirty(true); // force write
+        DumpBanlist();
+    }
+}
+
+BanMan::~BanMan()
+{
+    DumpBanlist();
+}
+
+void BanMan::DumpBanlist()
+{
+    SweepBanned(); // clean unused entries (if bantime has expired)
+
+    if (!BannedSetIsDirty())
+        return;
+
+    int64_t nStart = GetTimeMillis();
+
+    banmap_t banmap;
+    GetBanned(banmap);
+    if (m_ban_db.Write(banmap)) {
+        SetBannedSetDirty(false);
+    }
+
+    LogPrint(BCLog::NET, "Flushed %d banned node ips/subnets to banlist.dat  %dms\n",
+        banmap.size(), GetTimeMillis() - nStart);
+}
+
+void BanMan::ClearBanned()
+{
+    {
+        LOCK(m_cs_banned);
+        m_banned.clear();
+        m_is_dirty = true;
+    }
+    DumpBanlist(); //store banlist to disk
+    if(m_client_interface)
+        m_client_interface->BannedListChanged();
+}
+
+bool BanMan::IsBanned(CNetAddr ip)
+{
+    LOCK(m_cs_banned);
+    for (const auto& it : m_banned) {
+        CSubNet subNet = it.first;
+        CBanEntry banEntry = it.second;
+
+        if (subNet.Match(ip) && GetTime() < banEntry.nBanUntil) {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool BanMan::IsBanned(CSubNet subnet)
+{
+    LOCK(m_cs_banned);
+    banmap_t::iterator i = m_banned.find(subnet);
+    if (i != m_banned.end())
+    {
+        CBanEntry banEntry = (*i).second;
+        if (GetTime() < banEntry.nBanUntil) {
+            return true;
+        }
+    }
+    return false;
+}
+
+void BanMan::Ban(const CNetAddr& addr, const BanReason &banReason, int64_t bantimeoffset, bool sinceUnixEpoch) {
+    CSubNet subNet(addr);
+    Ban(subNet, banReason, bantimeoffset, sinceUnixEpoch);
+}
+
+void BanMan::Ban(const CSubNet& subNet, const BanReason &banReason, int64_t bantimeoffset, bool sinceUnixEpoch) {
+    CBanEntry banEntry(GetTime());
+    banEntry.banReason = banReason;
+    if (bantimeoffset <= 0)
+    {
+        bantimeoffset = m_default_ban_time;
+        sinceUnixEpoch = false;
+    }
+    banEntry.nBanUntil = (sinceUnixEpoch ? 0 : GetTime() )+bantimeoffset;
+
+    {
+        LOCK(m_cs_banned);
+        if (m_banned[subNet].nBanUntil < banEntry.nBanUntil) {
+            m_banned[subNet] = banEntry;
+            m_is_dirty = true;
+        }
+        else
+            return;
+    }
+    if(m_client_interface)
+        m_client_interface->BannedListChanged();
+
+    if(banReason == BanReasonManuallyAdded)
+        DumpBanlist(); //store banlist to disk immediately if user requested ban
+}
+
+bool BanMan::Unban(const CNetAddr &addr) {
+    CSubNet subNet(addr);
+    return Unban(subNet);
+}
+
+bool BanMan::Unban(const CSubNet &subNet) {
+    {
+        LOCK(m_cs_banned);
+        if (!m_banned.erase(subNet))
```

Comment:
Nit: `if (m_banned.erase(subNet) == 0)` to increase readability and avoid implicit conversion?